spinoso-string/src/enc/utf8/mod.rs

```diff
@@ -0,0 +1,1060 @@
+            Some(idx) => idx,
+        };
+        let mut slice = &self.inner[consumed..];
+        // Count of "characters" remaining until the `index`th character.
+        let mut remaining = index - consumed;
+        // This loop will terminate when either:
+        //
+        // - It counts `index` number of characters.
+        // - It consumes the entire slice when scanning for the
+        //   `index`th character.
+        //
+        // The loop will advance by at least one byte every iteration.
+        loop {
+            match bstr::decode_utf8(slice) {
+                // If we've run out of slice while trying to find the
+                // `index`th character, the lookup fails and we return `nil`.
+                (_, 0) => return None,
+
+                // The next two arms mean we've reached the `index`th
+                // character. Either return the next valid UTF-8
+                // character byte slice or, if the next bytes are an
+                // invalid UTF-8 sequence, the next byte.
+                (Some(_), size) if remaining == 0 => return Some(&slice[..size]),
+                // Size is guaranteed to be positive per the first arm
+                // which means this slice operation will not panic.
+                (None, _) if remaining == 0 => return Some(&slice[..1]),
+
+                // We found a single UTF-8 encoded characterk keep track
+                // of the count and advance the substring to continue
+                // decoding.
+                (Some(_), size) => {
+                    slice = &slice[size..];
+                    remaining -= 1;
+                }
+
+                // The next two arms handle the case where we have
+                // encountered an invalid UTF-8 byte sequence.
+                //
+                // In this case, `decode_utf8` will return slices whose
+                // length is `1..=3`. The length of this slice is the
+                // number of "characters" we can advance the loop by.
+                //
+                // If the invalid UTF-8 sequence contains more bytes
+                // than we have remaining to get to the `index`th char,
+                // then the target character is inside the invalid UTF-8
+                // sequence.
+                (None, size) if remaining < size => return Some(&slice[remaining..=remaining]),
+                // If there are more characters remaining than the number
+                // of bytes yielded in the invalid UTF-8 byte sequence,
+                // count `size` bytes and advance the slice to continue
+                // decoding.
+                (None, size) => {
+                    slice = &slice[size..];
+                    remaining -= size;
+                }
+            }
+        }
+    }
+
+    #[inline]
+    #[must_use]
+    pub fn get_char_slice(&self, range: Range<usize>) -> Option<&'_ [u8]> {
+        let Range { start, end } = range;
```

Comment:
can we add a TODO that we can do the same fast path check as line 222 if `start > self.len()`?